File: services/patient-service/src/controller/Memo.ts
import { Router, Request, Response } from "express";
import multer from "multer";
import * as MemoService from "../service/Memo";
import { requireAuth } from "../middleware/auth";

const router = Router({ mergeParams: true });

const ALLOWED_MIME_TYPES = new Set([
    "image/jpeg", "image/png",
    "application/pdf",
    "application/msword",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
]);

const upload = multer({
    storage: multer.memoryStorage(),
    limits: { fileSize: 10 * 1024 * 1024 }, // 10 MB
    fileFilter: (_req, file, cb) => {
        if (ALLOWED_MIME_TYPES.has(file.mimetype)) {
            cb(null, true);
        } else {
            cb(new Error(`File type not allowed: ${file.mimetype}`));
        }
    },
});

// GET /api/patients/:id/memos
router.get("/", requireAuth, async (req: Request, res: Response) => {
    const callerId = (req as any).callerId as string;
    if (callerId !== req.params.id) return res.status(403).json({ error: "Forbidden" });

    try {
        const memos = await MemoService.getMemos(req.params.id);
        res.json(memos);
    } catch (e) {
        console.error("[Memo] getMemos error:", e);
        res.status(500).json({ error: "Internal server error" });
    }
});

// POST /api/patients/:id/memos — text note
router.post("/", requireAuth, async (req: Request, res: Response) => {
    const callerId = (req as any).callerId as string;
    if (callerId !== req.params.id) return res.status(403).json({ error: "Forbidden" });

    const { title, content } = req.body;
    if (!title || !content) return res.status(400).json({ error: "title and content are required" });

    try {
        const memo = await MemoService.createTextMemo(req.params.id, title, content);
        res.status(201).json(memo);
    } catch (e) {
        console.error("[Memo] createTextMemo error:", e);
        res.status(500).json({ error: "Internal server error" });
    }
});

// POST /api/patients/:id/memos/upload — file upload
router.post("/upload", requireAuth, (req: Request, res: Response, next) => {
    upload.single("file")(req, res, (err) => {
        if (err instanceof multer.MulterError && err.code === "LIMIT_FILE_SIZE") {
            return res.status(400).json({ error: "File exceeds 10 MB limit" });
        }
        if (err) return res.status(400).json({ error: err.message });
        next();
    });
}, async (req: Request, res: Response) => {
    const callerId = (req as any).callerId as string;
    if (callerId !== req.params.id) return res.status(403).json({ error: "Forbidden" });

    if (!req.file) return res.status(400).json({ error: "file is required" });
    const { title } = req.body;
    if (!title) return res.status(400).json({ error: "title is required" });

    try {
        const memo = await MemoService.createFileMemo(req.params.id, title, req.file);
        res.status(201).json(memo);
    } catch (e) {
        console.error("[Memo] createFileMemo error:", e);
        res.status(500).json({ error: "Internal server error" });
    }
});

// POST /api/patients/:id/memos/doctor — called by consultation composite to store MC/prescription
// No ownership check — caller must be authenticated (JWT verified), patient_id comes from path
router.post("/doctor", requireAuth, async (req: Request, res: Response) => {
    const title = req.body.title as string;
    const content = req.body.content as string;
    const record_type = req.body.record_type as string;
    const issued_by = req.body.issued_by as string;

    if (!title || !content || !record_type || !issued_by) {
        return res.status(400).json({ error: "title, content, record_type, and issued_by are required" });
    }
    if (!["mc", "prescription"].includes(record_type)) {
        return res.status(400).json({ error: "record_type must be 'mc' or 'prescription'" });
    }

    try {
        const memo = await MemoService.createDoctorRecord(String(req.params.id), title, content, record_type as "mc" | "prescription", String(issued_by));
        res.status(201).json(memo);
    } catch (e) {
        console.error("[Memo] createDoctorRecord error:", e);
        res.status(500).json({ error: "Internal server error" });
    }
});

export default router;

File: composite/patient-orchestrator/src/routes/memo.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form
from src.models.memo import CreateTextMemoRequest, CreateDoctorRecordRequest, MemoResponse
from src.controller import memo as memo_controller
from src.dependencies import require_auth, AuthContext

router = APIRouter(prefix="/api/composite/patients", tags=["memos"])


@router.get("/{patient_id}/memos", response_model=list[MemoResponse])
async def get_memos(patient_id: str, auth: AuthContext = Depends(require_auth)):
    if auth.user_id != patient_id:
        raise HTTPException(status_code=403, detail="Forbidden")
    return await memo_controller.get_memos(patient_id)


@router.post("/{patient_id}/memos", response_model=MemoResponse, status_code=201)
async def create_text_memo(patient_id: str, body: CreateTextMemoRequest, auth: AuthContext = Depends(require_auth)):
    if auth.user_id != patient_id:
        raise HTTPException(status_code=403, detail="Forbidden")
    return await memo_controller.create_text_memo(patient_id, body)


@router.post("/{patient_id}/memos/upload", response_model=MemoResponse, status_code=201)
async def create_file_memo(
    patient_id: str,
    title: str = Form(...),
    file: UploadFile = File(...),
    auth: AuthContext = Depends(require_auth),
):
    if auth.user_id != patient_id:
        raise HTTPException(status_code=403, detail="Forbidden")
    return await memo_controller.create_file_memo(patient_id, title, file)


@router.post("/{patient_id}/memos/doctor", response_model=MemoResponse, status_code=201)
async def create_doctor_record(patient_id: str, body: CreateDoctorRecordRequest, auth: AuthContext = Depends(require_auth)):
    return await memo_controller.create_doctor_record(patient_id, body)
